spec.md

```diff
@@ -1090,6 +1119,287 @@ message ControllerServiceCapability {
+  // users to pass through the required credentials. This information is
+  // sensitive and MUST be treated as such (not logged, etc.) by the CO.
+  // This field is OPTIONAL.
+  map<string, string> controller_create_credentials = 4;
+
+  // NOTE: In the future, we may want to add “tags” to identify a
+  // snapshot. This is generated by CO internally and passed to the
+  // cloud storage backend through the plugin. This can serve the
+  // following purpose: Allow cloud storage admin to identify snapshots
+  // created by a specific CO.
+  // map<string, string> tags = 5;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
+  // The restored size of the snapshot in bytes.  This indicates how
+  // much data was written on the snapshot. It’s the minimum required
+  // size when creating a volume from snapshot. This field is OPTIONAL.
+  // If not set (value of 0), it indicates this size is unknown.
+  // The value of this field MUST NOT be negative.
+  int64 restored_size = 1;
+
+  // The size of the snapshot in bytes. This indicates the given size of
+  // the snapshot when it was created. This may be equal to or greater
+  // than the restored_size. This field is OPTIONAL. If not set (value
+  // of 0), it indicates this size is unknown.
+  // The value of this field MUST NOT be negative.
+  int64 size = 2;
+
+  // Contains identity information for the created snapshot. This field
+  // is REQUIRED. The identity information will be used by the CO in
+  // subsequent calls to refer to the provisioned snapshot.
+  string id = 3;
+
+  // Description for the created snapshot. This field is OPTIONAL.
+  string description = 4;
+
+  // Identity information for the source volume. Note that create
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string volume_id = 5;
+}
+```
+
+##### CreateSnapshot Errors
+
+If the plugin is unable to complete the CreateSnapshot call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Snapshot already exists but is incompatible | 6 ALREADY_EXISTS | Indicates that a snapshot corresponding to the specified snapshot `name` already exists but is incompatible with the specified `volume_id`. | Caller MUST fix the arguments or use a different `name` before retrying. |
```

Comment:
Is there a specific error code for "not enough space for a snapshot"

Reply:
grpc resource-exhausted ?